### ISSUE TITLE
Feature request: add more options to Luong attention

### ISSUE THREAD

**Issue by danielwatson6** [tensorflow/tensorflow#10842]
Currently, the best results of the Luong attention paper [(Minh-Thang Luong, Hieu Pham, Christopher D. Manning. "Effective Approaches to Attention-based Neural Machine Translation." EMNLP 2015.)](https://arxiv.org/abs/1508.04025) cannot be reproduced with the implementation of `tf.contrib.seq2seq.LuongAttention`. Features that are missing:

- **Local attention**: attend to a window of time steps, rather than to all of the time steps of the encoder output (global attention). The window size should be a hyperparameter that the user can tune.
- **Different scoring functions**: currently, the scoring function is limited to a dot products between each encoder output and the decoder output. The paper shows better results with "general" scoring (all of the encoder outputs are multiplied by one learnable matrix) and also explores the option of using Bahdanau-like scoring (concatenate and multiply by a learnable matrix, then apply tanh and take a dot product with a learnable vector).
- **Predictive alignments**: while the probability function can be replaced, it would be nice to add predictive alignment as a function, and make the implementation of both monotonic and predictive alignments behave well with local attention limited to a time window (changes shape of learnables).
- **Input-feeding approach**: (please correct me in this one if I am mistaken) the current implementation is missing the final step that computes a prediction by concatenating the context vector with the decoder output, weights them and applies tanh (let this be `s_t=tanh(W [c_t; h_t])`). Passing `s_t` through a softmax layer gives the prediction distribution, but passing it as is to the next input improved performance in the paper.

On a sidenote: it also seems to be that there is no difference in the key and query vectors between the Bahdanau and Luong attention mechanisms, when there should be. Bahdanau attention has a computation pathway starting from the previous decoder output `h_{t-1} -> a_t -> c_t -> h_t`, while Luong attention starts from the current output `h_t -> a_t -> c_t -> s_t`.

**Comment by aselle** [comment]
@ebrevdo. Is anyone going to work on this? Should we mark this contributions welcome if we want this included but won't do it at google.

**Comment by ebrevdo** [comment]
Thang, can you comment?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
@lmthang which of these features do you feel is important?

**Comment by zhouyangnk** [comment]
@ebrevdo  Local attention, I think

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @lmthang, @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by lmthang** [comment]
Hi all,

Sorry for a late reply. Here's my brief answer:
* Local attention, predictive alignments: unfortunately, we don't have the bandwidth for this. Local attention, while being an interesting idea, is more complicated to implement and train. Contributions are welcome though I expect the contributors need to demonstrate convincing gains on over nowadays strong [baselines](github.com/tensorflow/nmt) and new [advances in attention](https://arxiv.org/abs/1706.03762).

* Input-feeding approach: yes, we missed the tanh() part but the performance is found to be similar. See answers in  https://github.com/tensorflow/nmt/issues/257.

* Different scoring functions: in fact, we are implementing the general approach for Luong attention, see this [line](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py#L402). Similarly for Bahdanau attention if you read the [code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/seq2seq/python/ops/attention_wrapper.py#L546) carefully. 

Hope that clarifies!

**Comment by tensorflowbutler** [comment]
Nagging Assignees @lmthang, @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @lmthang, @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @lmthang, @ebrevdo: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @lmthang, @ebrevdo: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @lmthang, @ebrevdo: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
